src/coreComponents/mesh/generators/ParticleBlockABC.hpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2020-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+#ifndef GEOSX_PARTICLEBLOCKABC_HPP
+#define GEOSX_PARTICLEBLOCKABC_HPP
+
+#include "dataRepository/Group.hpp"
+#include "mesh/ParticleType.hpp"
+#include "common/DataTypes.hpp"
+
+#include <vector>
+
+namespace geosx
+{
+
+/**
+ * Abstract base class defining the information provided by any particle block implementation.
+ * Mainly particle type-related information (TODO put an example here),
+ * a local to global mapping and some kind of accessors to external properties.
+ *
+ * The particles of ParticleBlockABC are all of the same kind.
+ *
+ * It's noteworthy that the ParticleBlockABC is immutable oriented.
+ * The derived implementations need to have the modification/creation capabilities.
+ */
+class ParticleBlockABC : public dataRepository::Group
+{
+public:
+
+  /**
+   * @brief Constructor
+   * @param name The name of this Group.
+   * @param parent The parent Group.
+   */
+  ParticleBlockABC( string const & name,
+                    Group * const parent )
+    :
+    Group( name, parent )
+  { }
+
+  /**
+   * @brief Get the type of particle in this subregion.
+   * @return a string specifying the type of particle in this subregion
+   *
+   * See class FiniteParticleBase for possible particle type.
+   */
+  virtual ParticleType getParticleType() const = 0;
+
+  virtual array1d< int > getParticleID() const = 0;
+
+  virtual void setParticleID(array1d< int > const particleID) = 0;
```

Comment:
In the current pattern with the `*ABC` classes, the abstract base class are meant to be consumed by some client code. Seeing setters here mean that your pattern is different. What was your need?
Also I spot no `ParticleBlockManagerABC` so you expose the implementation?

Reply:
I don't know enough about this to comment, I was just copying the architecture for the cell block manager. I did away with `ParticleBlockManagerABC` since there is currently only one kind of particle block. Should this be addressed now or fixed in a future PR?

Reply:
The idea is to have an abstraction (say `ParticleBlockManagerABC` like for `CellBlockManagerABC`) that is used outside of the "meshing" zone.

Reply:
I looked at `CellBlockManagerABC` again and now I remember why I didn't make a `ParticleBlockManagerABC` - `CellBlockManagerABC` doesn't contain any functions or variables that make sense for particles. That being said, I'll implement it anyways since it's possible something may go there someday, but for now it will be basically empty.